vernac/g_vernac.mlg

```diff
@@ -654,13 +654,15 @@ GRAMMAR EXTEND Gram
           LIST1 [ v=strategy_level; "["; q=LIST1 smart_global; "]" -> { (v,q) } ] ->
             { VernacSetStrategy l }
       (* Canonical structure *)
-      | IDENT "Canonical"; IDENT "Structure"; qid = global ->
-          { VernacCanonical CAst.(make ~loc @@ AN qid) }
+      | IDENT "Canonical"; IDENT "Structure"; qid = global; ud = OPT [ u = OPT univ_decl; d = def_body -> { (u,d) } ] ->
+          { match ud with
+           | None ->
+             VernacCanonical CAst.(make ~loc @@ AN qid)
+           | Some (u,d) ->
+             let s = coerce_reference_to_id qid in
+             VernacDefinition ((NoDischarge,CanonicalStructure),((CAst.make (Name s)),u),d) }
       | IDENT "Canonical"; IDENT "Structure"; ntn = by_notation ->
           { VernacCanonical CAst.(make ~loc @@ ByNotation ntn) }
-      | IDENT "Canonical"; IDENT "Structure"; qid = global; u = OPT univ_decl; d = def_body ->
-          { let s = coerce_reference_to_id qid in
-          VernacDefinition ((NoDischarge,CanonicalStructure),((CAst.make (Name s)),u),d) }
 
       (* Coercions *)
       | IDENT "Coercion"; qid = global; u = OPT univ_decl; d = def_body ->
```

Comment:
You probably need to do the same for Coercion too.

Reply:
The problem in you original commit is that  camlp5 is not able to factor line 661 with 657, dunno why, so I factored them by hand. For coercions there is only one line, so no factoring seems needed

Reply:
Feel free to push (more) tests on my branch

Reply:
It doesn't need manual factoring with `IDENT "Coercion"; qid = global; ":"; s = class_rawexpr`?

Reply:
I don't think this command needs universe levels, it only declares a posteriori a coercion.

Reply:
>I don't think this command needs universe levels, it only declares a posteriori a coercion.

`Coercion foo := term.` defines foo so should allow universe binders.

Reply:
Ok, I've added a test but there was no factoring needed since there were not 2 conflicting grammar rules for that.